### PR TITLE
Mirror service setup with OSV implementation

### PR DESCRIPTION
**Pull request by sahibamittal**
https://github.com/mehab/DTKafkaPOC/issues/55

This PR includes basic setup of new module mirror-service as well as migration of OSV mirroring as an event-triggered task using Kafka streams.

NOTE: Currently, I've used OSV advisory as common data schema to publish on topic "dtrack.vulnerability". This would be replaced by a different (TBD) schema later.

### CODE REVIEW

File: mirror-service/src/main/java/org/acme/client/OsvClient.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.Optional;
+
+import static org.acme.util.FileUtil.getTempFileLocation;
+
+/**
+ * Client for the Snyk REST API.
+ * <p>
+ * Note: Using the MicroProfile RestClient does not work, because Snyk expects the PURL
+ * to be encoded differently than what the MP RestClient is doing (it expects ":" and "@"
+ * to be encoded, which is normally not necessary for URL path segments).
```
Comment (nscuro) [inline mirror-service/src/main/java/org/acme/client/OsvClient.java:30]: Does not apply here, as this was specific to the Snyk client.

File: mirror-service/src/test/java/org/acme/client/OsvClientTest.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        wireMockServer.resetAll();
+    }
+
+    @Test
+    void testDownloadOsvZips() throws Exception {
+        // Have the OSV API return one zip with one file for the provided ecosystem
+        wireMockServer.stubFor(get(anyUrl())
+                .willReturn(aResponse()
+                        .withStatus(HttpStatus.SC_OK)
+                        .withResponseBody(fromJsonBytes(getOsvTestZipData("osv-vuln.json")))));
```
Comment (nscuro) [inline mirror-service/src/test/java/org/acme/client/OsvClientTest.java:60]: `fromJsonBytes` looks odd given this is a ZIP file. Is there an alternative method that could be used?

File: mirror-service/src/main/java/org/acme/osv/OsvAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+import static org.acme.util.FileUtil.deleteFileAndDir;
+
+@ApplicationScoped
+public class OsvAnalyzer {
```
Comment (nscuro) [inline mirror-service/src/main/java/org/acme/osv/OsvAnalyzer.java:25]: This class should be renamed to be more reflective of what it does. It is not analyzing anything, so `Analyzer` may not be the right name.

File: mirror-service/src/test/java/org/acme/osv/OsvAnalyzerTest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        Path tempZipLocation = getTempFileLocation("test", ".zip");
+        Files.copy(testFile, tempZipLocation, StandardCopyOption.REPLACE_EXISTING);
+        when(osvClientMock.downloadEcosystemZip(anyString()))
+                .thenReturn(tempZipLocation);
+        inputTopic.pipeInput("Go", "");
+        assertThat(outputTopic.getQueueSize()).isEqualTo(1);
+        assertThat(outputTopic.readRecord()).satisfies(record -> {
+            assertThat(record.key()).isEqualTo("OSV/GO-2020-0023");
+        });
+        deleteFileAndDir(tempZipLocation);
```
Comment (nscuro) [inline mirror-service/src/test/java/org/acme/osv/OsvAnalyzerTest.java:74]: This should be in a `@AfterEach` method, to ensure that it's always called, regardless whether the test failed or not.

File: docker-compose.yml [hunk 5]
```diff
@@ -70,6 +70,26 @@ services:
...
+    depends_on:
+      - postgres
+      - redpanda
+    environment:
+      QUARKUS_KAFKA_STREAMS_BOOTSTRAP_SERVERS: "dt-redpanda:29092"
+      KAFKA_STREAMS_NUM_STREAM_THREADS: "1"
+      # KAFKA_STREAMS_NUM_STREAM_THREADS: "3"
+      QUARKUS_DATASOURCE_JDBC_URL: "jdbc:postgresql://dt-postgres:5432/dtrack"
+      QUARKUS_DATASOURCE_USERNAME: "dtrack"
+      QUARKUS_DATASOURCE_PASSWORD: "dtrack"
```
[the remark this answers is not in the document]
Comment (sahibamittal) [reply docker-compose.yml:84]: Mirror-service should not need to access database, I'll remove these props.

**Review by nscuro — APPROVED** [review]

**Pull request merged**